plugins/inventory/terraform_state.py

```diff
@@ -0,0 +1,246 @@
+    type: bool
+    default: false
+  binary_path:
+    description:
+      - The path of a terraform binary to use.
+    type: path
+  hostnames:
+    description:
+      - A list in order of precedence for hostname variables.
+      - The elements of the list can be a dict with the keys mentioned below or a string.
+      - Can be one of the options specified in U(https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/instance#argument-reference).
+      - If value provided does not exist in the above options, it will be used as a literal string.
+      - To use tags as hostnames use the syntax tag:Name=Value to use the hostname Name_Value, or tag:Name to use the value of the Name tag.
+      - If not provided the final hostname will be C(terraform resource type) + C(_) + C(terraform resource name)
+    type: list
+    elements: raw
+    default: []
+    suboptions:
+      name:
+        description:
+          - Name of the host.
+        type: str
+        required: True
+      prefix:
+        description:
+          - Prefix to prepend to I(name). Same options as I(name).
+          - If I(prefix) is specified, final hostname will be I(prefix) +  I(separator) + I(name).
+        type: str
+        default: ''
+        required: False
+      separator:
+        description:
+          - Value to separate I(prefix) and I(name) when I(prefix) is specified.
+        type: str
+        default: '_'
+        required: False
+"""
+
+EXAMPLES = r"""
+# Example using constructed features to set ansible_host
+plugin: cloud.terraform.terraform_state
+backend_config: |
+  backend "http" {
+      address = "https://localhost:8043/api/v2/state/3/"
+      skip_cert_verification = true
+      username = "ansible"
+      password = "test123!"
+  }
+compose:
+  # Use the public IP address to connect to the host
+  ansible_host: public_ip
+"""
+
+
+import os
+from tempfile import TemporaryDirectory
+from typing import Any, Dict, List, Optional
+
+from ansible.module_utils._text import to_text
+from ansible.module_utils.common import process
+from ansible.plugins.inventory import Constructable
+from ansible_collections.cloud.terraform.plugins.module_utils.errors import TerraformError, TerraformWarning
+from ansible_collections.cloud.terraform.plugins.module_utils.models import TerraformModuleResource
+from ansible_collections.cloud.terraform.plugins.module_utils.terraform_commands import TerraformCommands
+from ansible_collections.cloud.terraform.plugins.module_utils.utils import validate_bin_path
+from ansible_collections.cloud.terraform.plugins.plugin_utils.base import TerraformInventoryPluginBase
+from ansible_collections.cloud.terraform.plugins.plugin_utils.common import module_run_command
+
+
+def filter_instances(resources: List[TerraformModuleResource], types: List[str]) -> List[TerraformModuleResource]:
+    # should we add additional filtering on provider (provider_name='registry.terraform.io/hashicorp/aws') ?
```

Comment:
We should use provider_name + type here. Generally, it should be ok with just type, but there may be some edge cases where it could be a problem. Safer to just go with the unambiguous combination of provider and type.

Reply:
done